### PR TITLE
Refactor expected exception recipe - fixes #72

### LINKED ISSUE #72: ExpectedException to assertThrows does not convert when parameter is hamcrest matcher
discovered while applying `org.openrewrite.java.spring.Boot2JUnit4to5Migration` to spring-cloud-consul
https://github.com/spring-cloud/spring-cloud-consul/blob/1da019c336fee08be42347e6f17a8664e81045e1/spring-cloud-consul-binder/src/test/java/org/springframework/cloud/consul/binder/config/ConsulBinderConfigurationTests.java#L39-L48
```
...
import static org.hamcrest.Matchers.containsString;
...
public class ConsulBinderConfigurationTests {
@Test
@Disabled
    void consulBinderDisabledWorks() {
		this.exception.expectMessage(containsString("no proper implementation found"));
		new SpringApplicationBuilder(Application.class).initializers(new ConsulTestcontainers())
				.properties("spring.cloud.consul.binder.enabled=false").run();
	}
```

### PR DESCRIPTION
**Pull request by traceyyoshima**
Refactor the recipe to work with general uses of the ExpectedException class.
fixes #72

### CODE REVIEW

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 1]
```diff
@@ -32,7 +33,19 @@
 import java.util.stream.Collectors;
 
 /**
- * Replace usages of JUnit 4's @Rule ExpectedException with JUnit 5 Assertions.assertThrows
+ * Replace usages of JUnit 4's @Rule ExpectedException with JUnit 5 Assertions.
+ *
+ * The recipe refactors of ExpectedException to org.junit.jupiter.api.assertThrows,
+ * and org.junit.jupiter.api.assertThrows when applicable.
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:39]: This sentence is a bit confusing and repeats `org.junit.jupiter.api.assertThrows` twice. Maybe just delete this sentence?

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 2]
```diff
@@ -41,16 +54,17 @@
...
                             "import org.junit.jupiter.api.function.Executable;" +
-                            "class AssertThrows {\n" +
-                            "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,Supplier<String> messageSupplier){}" +
-                            "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,String message){}" +
-                            "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable){}" +
-                            "}"),
+                            "class AssertThrows {" +
+                                "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,Supplier<String> messageSupplier){}" +
+                                "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,String message){}" +
+                                "static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable){}" +
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:60]: I know the previous example didn't have this either, but add a return null; to each method body will prevent compiler errors when parsing. These are silent, but we have seen some performance impact when we get a large number of these.

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 3]
```diff
@@ -93,59 +116,128 @@ public String getDescription() {
...
+                    if (!isExpectArgAMatcher) {
+                        return m;
+                    }
+                } else {
+                    final JavaType.FullyQualified argType = TypeUtils.asFullyQualified(expectMethodArg.getType());
+                    if (argType == null || !argType.getFullyQualifiedName().equals("java.lang.Class")) {
+                        return m;
+                    }
+                }
+            }
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:147]: In the above code, we return out of the method when the method invocation does not match what we are expecting. This will short-circuit any potential checks for the other two method invocations.

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 4]
```diff
@@ -93,59 +116,128 @@ public String getDescription() {
...
-                                            .build(),
-                                    m.getBody().getCoordinates().replace(),
-                                    expectedException, printedStatements.toString(), expectedMessage));
+
+                final Expression expectCauseMethodArg = args.get(0);
+                if (expectCauseMethodArg instanceof J.MethodInvocation) {
+                    isExpectedCauseArgAMatcher = isHamcrestMatcher((J.MethodInvocation) expectCauseMethodArg);
+                    if (!isExpectedCauseArgAMatcher) {
+                        return m;
+                    }
```
Comment (tkvangorder) [inline src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:182]: Same story here, if things don't match up, we don't execute the template, maybe this is ok and by design? just wanted to point it out.

**Review by tkvangorder — CHANGES_REQUESTED** [review]

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java (hunk 3 above)
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:144]: Ya, it's designed to ensure the method arguments match the recipe expectations prior to making changes.
The ExpectedException methods args have remained the same, since it was added in '12.
So, technically, the recipe will refactor the current JUnit4 version and previous versions, but it's possible for changes to happen eventually.

I.E. there is an open issue to remove the dependency on Hamcrest matcher - they tried to deprecate it, but the deprecation was reverted.

I'm not sure how to handle those types of edge cases and couldn't find a precedence for future proofing recipes.
If you think it's worth while, I could add code to leave behind an artifact like a comment about why the recipe didn't refactor the test.

Comment (tkvangorder) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java:144]: I am good with how it works now, its much better than it was.

**Review by sambsnyd — APPROVED** [review]
Good work

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 5]
```diff
@@ -93,59 +116,128 @@ public String getDescription() {
...
-                        .collect(Collectors.toList());
+                final Expression expectMessageMethodArg = args.get(0);
+                if (expectMessageMethodArg instanceof J.MethodInvocation) {
+                    isExpectMessageArgAMatcher = isHamcrestMatcher((J.MethodInvocation) expectMessageMethodArg);
+                    if (!isExpectMessageArgAMatcher) {
+                        return m;
+                    }
+                } else {
+                    if (!(expectMessageMethodArg instanceof J.Literal &&
+                            expectMessageMethodArg.getType() == JavaType.Primitive.String)) {
```
[the remark this answers is not in the document]
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java]: Thanks - updated

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java [hunk 6]
```diff
@@ -93,59 +116,128 @@ public String getDescription() {
...
-                                            .javaParser(ASSERTIONS_PARSER.get())
-                                            .staticImports("org.junit.jupiter.api.Assertions.assertThrows")
-                                            .build(),
-                                    m.getBody().getCoordinates().replace(),
-                                    expectedException, printedStatements.toString(), expectedMessage));
+
+                final Expression expectCauseMethodArg = args.get(0);
+                if (expectCauseMethodArg instanceof J.MethodInvocation) {
+                    isExpectedCauseArgAMatcher = isHamcrestMatcher((J.MethodInvocation) expectCauseMethodArg);
+                    if (!isExpectedCauseArgAMatcher) {
```
[the remark this answers is not in the document]
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java]: Updated the method and simplified the conditions.

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java (hunk 1 above)
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java]: Removed

File: src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java (hunk 2 above)
Comment (traceyyoshima) [reply src/main/java/org/openrewrite/java/testing/junit5/ExpectedExceptionToAssertThrows.java]: Added the returns

**Review by tkvangorder — APPROVED** [review]
Looks great!
